easybuild/tools/containers.py

```diff
@@ -0,0 +1,304 @@
+    else:
+        raise EasyBuildError("--container-base must be specified")
+
+    # first argument to --container-base is the Singularity bootstrap agent (localimage, shub, docker)
+    bootstrap_agent = base_specs[0]
+
+    # check bootstrap type value and ensure it is localimage, shub, docker
+    if bootstrap_agent not in SINGULARITY_BOOTSTRAP_TYPES:
+        known_bootstrap_agents = ', '.join(SINGULARITY_BOOTSTRAP_TYPES)
+        raise EasyBuildError("Bootstrap agent in container base spec must be one of: %s" % known_bootstrap_agents)
+
+    res = {'bootstrap_agent': bootstrap_agent}
+
+    for idx, base_spec in enumerate(base_specs[1:]):
+        res.update({'arg%d' % (idx + 1): base_spec})
+
+    return res
+
+
+def generate_singularity_recipe(easyconfigs, container_base):
+    """Main function to Singularity definition file and image."""
+
+    cont_path = container_path()
+
+    # check if --containerpath is valid path and a directory
+    if os.path.isdir(cont_path):
+        _log.info("Path for container recipes & images: %s", cont_path)
+    else:
+        raise EasyBuildError("Location for container recipes & images is a non-existing directory: %s" % cont_path)
+
+    base_specs = parse_container_base(container_base)
+
+    # extracting application name,version, version suffix, toolchain name, toolchain version from
+    # easyconfig class
+
+    bootstrap_agent = base_specs['bootstrap_agent']
+
+    base_image, base_image_tag = None, None
+
+    # with localimage it only takes 2 arguments. --container-base localimage:/path/to/image
+    # checking if path to image is valid and verify image extension is '.img' or '.simg'
+    if base_specs['bootstrap_agent'] == LOCALIMAGE:
+        base_image = base_specs['arg1']
+        if os.path.exists(base_image):
+            # get the extension of container image
+            image_ext = os.path.splitext(base_image)[1]
+            if image_ext == '.img' or image_ext == '.simg':
+                _log.debug("Extension for base container image to use is OK: %s", image_ext)
+            else:
+                raise EasyBuildError("Invalid image extension '%s' must be .img or .simg", image_ext)
+        else:
+            raise EasyBuildError("Singularity base image at specified path does not exist: %s", base_image)
+
+    # otherwise, bootstrap agent is 'docker' or 'shub'
+    # format --container-base {docker|shub}:<image>:<tag>
+    else:
+        base_image = base_specs['arg1']
+        # image tag is optional
+        base_image_tag = base_specs.get('arg2', None)
+
+    bootstrap_from = base_image
+    if base_image_tag:
+        bootstrap_from += ':' + base_image_tag
+
+    # if there is osdependencies in easyconfig then add them to Singularity recipe
+    install_os_deps = ''
+    osdeps = easyconfigs[0]['ec']['osdependencies']
+    for osdep in osdeps:
+        if isinstance(osdep, basestring):
+            install_os_deps += "yum install -y %s\n" % osdep
```

Comment:
With the enhanced implementation, it now looks like this:

```
$ grep %post -A 1 Singularity.OpenMPI-2.1.0-GCC-6.3.0-2.28
%post
yum install -y libibverbs-dev || yum install -y libibverbs-devel || yum install -y rdma-core-devel
```

Which is more correct than what you had imho, i.e. if *one* of these is installed, it's fine. 

With your approach, it was possible multiple ones got installed, at least in theory.

Reply:
you can also `yum install --skip-broken install -y libibverbs-dev libibverbs-devel rdma-core-devel || true`. The final `true` is needed because it can be the case that all of the dependencies are not available. I don't know whether that's desirable or not though.

Reply:
i agree with @FooBarQuaxx should add ``|| true`` at the end in case none of the deps are not available.

Reply:
That was the thing, I don't know which approach makes more sense at that point. But I guess carrying on is kind of more optimistic.